src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/JNativeScanTask.java

```diff
@@ -0,0 +1,195 @@
+    private final List<Path> modulePaths;
+    private final List<String> cmdRootModules;
+    private final Runtime.Version version;
+    private final Action action;
+
+    public JNativeScanTask(PrintWriter out, List<Path> classPaths, List<Path> modulePaths,
+                           List<String> cmdRootModules, Runtime.Version version, Action action) {
+        this.out = out;
+        this.classPaths = classPaths;
+        this.modulePaths = modulePaths;
+        this.version = version;
+        this.action = action;
+        this.cmdRootModules = cmdRootModules;
+    }
+
+    public void run() throws JNativeScanFatalError {
+        List<ClassFileSource> toScan = new ArrayList<>(findAllClassPathJars());
+
+        ModuleFinder moduleFinder = ModuleFinder.of(modulePaths.toArray(Path[]::new));
+        List<String> rootModules = cmdRootModules;
+        if (rootModules.contains("ALL-MODULE-PATH")) {
+            rootModules = allModuleNames(moduleFinder);
+        }
+        Configuration config = Configuration.resolveAndBind(moduleFinder, List.of(systemConfiguration()),
+                ModuleFinder.of(), rootModules);
+        for (ResolvedModule m : config.modules()) {
+            toScan.add(new ClassFileSource.Module(m.reference()));
+        }
+
+        SortedMap<ClassFileSource, SortedMap<ClassDesc, List<RestrictedUse>>> allRestrictedMethods;
+        try(ClassResolver classesToScan = ClassResolver.forClassFileSources(toScan, version);
+            ClassResolver systemClassResolver = ClassResolver.forSystemModules(version)) {
+            NativeMethodFinder finder = NativeMethodFinder.create(classesToScan, systemClassResolver);
+            allRestrictedMethods = finder.findAll();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        switch (action) {
+            case PRINT -> printNativeAccess(allRestrictedMethods);
+            case DUMP_ALL -> dumpAll(allRestrictedMethods);
+        }
+    }
+
+    private List<ClassFileSource> findAllClassPathJars() throws JNativeScanFatalError {
+        List<ClassFileSource> result = new ArrayList<>();
+        for (Path path : classPaths) {
+            if (isJarFile(path)) {
+                Deque<Path> jarsToScan  = new ArrayDeque<>();
+                jarsToScan.offer(path);
+
+                // recursively look for all class path jars, starting at the root jars
+                // in this.classPaths, and recursively following all Class-Path manifest
+                // attributes
+                while (!jarsToScan.isEmpty()) {
+                    Path jar = jarsToScan.poll();
+                    String[] classPathAttribute = classPathAttribute(jar);
+                    Path parentDir = jar.getParent();
+                    for (String classPathEntry : classPathAttribute) {
+                        Path otherJar = parentDir != null
+                                ? parentDir.resolve(classPathEntry)
```

Comment:
We'lll need to create a follow on issue to re-examine this as the value of a Class-Path attribute aren't sequence of relative URIs (with an optional "file" URI scheme) rather than file paths. Treating it as a file path may work in some cases but won't work once you encounter cases that use escaping.

Reply:
Okay, it seems that I didn't read the spec careful enough, and filtering valid URLs in the Class-Path attribute is actually a bit tricky. Let's handle this in a follow up.